compiler/sem/mirtraverse.nim

```diff
@@ -0,0 +1,539 @@
+            state = external(T)
+
+          dec inLoop
+
+        states[instr.id] = state
+
+      dec pc
+      next =
+        if pc >= 0: c[pc].node
+        else: tree.len
+
+    yield (i, tree[i])
+
+    if retry:
+      assert inLoop > 0, "uncertainity may only occur inside loops"
+      # an uncertain state was observed, queue the node for retrying and
+      # remeber the code position
+      let pos = InstrPos(pc + 1)
+      if output.reevaluate.len == 0 or output.reevaluate[^1] != pos:
+        output.reevaluate.add pos
+
+      output.retry.add (i, output.reevaluate.high)
+      retry = false
+
+    dec i
+
+iterator processRetries*[T](tree: MirTree, span: Slice[Node],
+                            c: ControlFlowGraph, o: var Output[T]
+                           ): tuple[n: Node, state: T] =
+  # TODO: document
+  mixin external, uncertain, isUncertain, merge
+  if o.reevaluate.len > 0:
+    # propagate the input -> output state until the state for each element in
+    # `reevaluate` is not uncertain anymore
+    # XXX: this is inefficient. We only need to propagate the state of
+    #      the join-points that are connected to the instructions in the
+    #      `reevaluate` set
+    let
+      firstPc = first(c, span)
+      lastPc = last(c, span)
+    var p = 0 # position into `o.reevaluate`
+
+    template states: untyped = o.states
+
+    block propagate:
+      while true:
+        var active = external(T)
+
+        for pc in countdown(lastPc, firstPc):
+          case c[pc].op
+          of opGoto:
+            if c[c.map[c[pc].dest]].node in span:
+              active = states[c[pc].dest]
+            else:
+              active = external(T)
+
+          of opFork:
+            let src =
+              if c[c.map[c[pc].dest]].node in span:
+                states[c[pc].dest]
+              else:
+                external(T)
+
+            # TODO: document the reason behind this part
+            if not isUncertain(src):
+              # XXX: alternatively, ``merge`` could be *required* to be a
+              #      no-op if 'b' is the "uncertain" state
+              active.merge(src)
+
+          of opLoop:
+            active = states[c[pc].dest]
+          of opJoin:
+            let id = c[pc].id
+            if isUncertain(states[id]):
+              # update join points state. Note that in the case of some
+              # nested loops, it's possible and valid that `active` is also in
+              # the "uncertain" state. Further iterations are needed then
+              states[id] = active
+            else:
+              when defined(debug):
+                active.merge(states[id])
+                # the assertion failing can be caused by an incorrect ``merge``
+                # implementation
+                assert states[id] == active
+
+          # XXX: is it possible for an item coming later in `reevaluate` to
+          #      already be in a "certain" state while an earlier one is not?
```

Comment:
Only if there exist no dependencies between the two parts of the tree, right? If that's is not the case then we cannot guarantee convergence of the fixed point approach. Likely need some sort of assert but I'm not exactly sure how that might work.

I think a scenario where this might occur is in a loop, where the body behaviour is certain but the loop condition is not? I'm likely grossly misunderstanding the MIR translation and processing, though.

Reply:
Okay, so what the `traverseReverse` iterator does is to "execute" a span of the control-flow-graph in reverse, propagating a property to all join points.

The iterator itself only takes care of propagating the property and figuring out its state at the entry (or in this case exit) of a basic block -- evaluating said basic block is the responsibility of the iterator's user.

When starting execution, the state of each join point is set to a user-provided default state. For example, in the case of ownership analysis, this would be `yes`. The CFG instructions (which are "executed" before yielding a basic block) are evaluated as following:
- `goto(x) = state(x)`
- `fork(x, y) = union(state(x), state(y))`
- `loop(x) = uncertain`

The pseudo-function `state` either evalutes to the state of the given join-point, or, if the join-point lies outside the executed span, the user-provided default state.

If during the evaluation of the basic-block, the current state of the property is sampled and happens to be `uncertain`, the basic-block and the MIR node at which the sampling occurred are remembered.

The `processRetries` iterator is then responsible for propagating the state between join-points until the would-be-input to all basic-blocks that were remembered earlier is not `uncertain` anymore, yielding the now not-uncertain state together with the also remembered node position.

---

For the comment in question, I believe the answer is: "yes, it can happen". Consider this CFG:
```
0: join
fork 1 # <- `while` condition
fork 5
1: join

... # sample position B
2: join
fork 3 # <- `while` condition
fork 4
3: join
... # the state is set to "yes" here
loop 2
4: join # <- inner `while` end

... # sample position A
loop 0
5: join # <- outer `while` end
```
which would be generated for:
```nim
while cond_a:
  discard "sample position B"
  while cond_b:
    ... # the state is set to "yes" here

  discard "sample position A"
```

In this case, when propagation first reaches `B`, the state would be certain, while for `A` it would still be `uncertain`. This only affects the order in which the retry items are yielded however, which in all current use-cases doesn't matter.